File: ruby/robot-name/robot_name.rb
require 'set'

class Robot
  LETTERS = ('A'..'Z').to_a

  @@names = Set.new

  def name
    reset if @name.nil?
    @name
  end

  def reset
    @name = unique_name
  end

  private

  def name_exists?(name)
    @@names.include? name
  end

  def unique_name
    name = gen_name
    name = gen_name while name_exists?(name)
    @@names.add(name)
    name
  end

  def gen_name
    "#{prefix}#{designation}"
  end

  def designation
    format('%03d', rand(999))
  end

  def prefix
    "#{LETTERS.sample}#{LETTERS.sample}"
  end
end

File: clojure/robot-name/src/robot_name.clj
(ns robot-name)

(defn- char-range
  [start end]
  (map char (range (int start) (+ 1 (int end)))))

(def alphabet (char-range \A \Z))

(defn- prefix
  []
  (str (rand-nth alphabet) (rand-nth alphabet)))

(defn- serial
  []
  (format "%03d" (int (rand 999))))

(defn- gen-name
  []
  (str (prefix) (serial)))

(defn robot
  []
  {:name (gen-name)})

(defn reset-name
  [bot]
  (assoc bot :name (gen-name)))

(defn robot-name
  [bot]
  (bot :name))
